### PR TITLE
added handling of stream/message tags

### PR DESCRIPTION
**Pull request by RalphSteinhagen**
Added tag_t definition and basic stream-tag functionality to the port and node definitions.

```
/**
 * @brief 'ports' are interfaces that allows data to flow between blocks in a graph, similar to RF connectors.
 * Each block can have zero or more input/output ports. When connecting ports, either a single-step or a two-step
 * connection method can be used. Ports belong to a computing domain, such as CPU, GPU, or FPGA, and transitions
 * between domains require explicit data conversion.
 * Each port consists of a synchronous performance-optimised streaming and asynchronous tag communication component:
 *                                                                                      ┌───────────────────────
 *         ───────────────────┐                                       ┌─────────────────┤  <node/block definition>
 *             output-port    │                                       │    input-port   │  ...
 *          stream-buffer<T>  │>───────┬─────────────────┬───────────>│                 │
 *          tag-buffer<tag_t> │      tag#0             tag#1          │                 │
 *                            │                                       │                 │
 *         ───────────────────┘                                       └─────────────────┤
 *
 * Tags contain the index ID of the sending/receiving stream sample <T> they are attached to. Node implementations
 * may choose to chunk the data based on the MIN_SAMPLES/MAX_SAMPLES criteria only, or in addition break-up the stream
 * so that there is only one tag per scheduler iteration. Multiple tags on the same sample shall be merged to one.
 * [..]
 */
 ```

* added `tag_t` and some default tag key-value-type definitions (in-line with an earlier approvedd [GR 4.0](https://github.com/gnuradio/gnuradio/pull/6362) PRs)
* renamed the reader/writer to more explicit `streamReader`/`streamWriter` and `tagReader`/`tagWriter` in the public API to more explicitly identify if the buffer is intended to transmit samples of type <T> or tags of type <tag_t> (N.B. functionally these buffers are otherwise identical).

This implementation assumed the same buffer size for streams and tags since -- in theory -- one could attach as many tags as samples per processing iteration. N.B. for large buffers and large numbers of ports/blocks in the application this can be wasteful and one should find a good policy w.r.t. the usually rare occurrence of tags compared to samples. I.e. 10%|1%|0.1% of buffer size being allocated for tags?!?!

Missing corner cases/features:
* handling when the distance of the next tag is less than the port's MIN_SAMPLES (rare corner case implement later)
* convenience function of forwarding tags from input ports to output ports for those implementing their `work()` function from scratch

Tackles '[stream tag](https://wiki.gnuradio.org/index.php?title=Stream_Tags) handling' in https://github.com/fair-acc/opendigitizer/issues/46.

### CODE REVIEW

File: include/port.hpp [hunk 1]
```diff
@@ -304,6 +369,27 @@ static_assert(OUT_MSG<float, 0, 0, "out_msg">::static_name() == fixed_string("ou
...
+            },
+            1_UZ);
 }
 
+constexpr void
+publish_tag(Port auto &port, const tag_t::map_type &tag_data, std::size_t tag_offset = 0) noexcept {
+    port.tagWriter().publish(
+            [&port, &tag_data, &tag_offset](std::span<fair::graph::tag_t> tag_output) {
+                tag_output[0].index = port.streamWriter().position() + tag_offset;
+                tag_output[0].map.clear();
```
[the remark this answers is not in the document]
Comment (RalphSteinhagen) [reply include/port.hpp:387]: fixed to copy-constructor done

File: include/port.hpp [hunk 2]
```diff
@@ -196,60 +235,88 @@ class port {
...
+                _tagIoHandler = TagBufferType(min_size).new_writer();
             } catch (...) {
                 return connection_result_t::FAILED;
             }
         }
         return connection_result_t::SUCCESS;
     }
 
-    [[nodiscard]] BufferType
+    [[nodiscard]] std::pair<BufferType, TagBufferType>
```
[the remark this answers is not in the document]
Comment (RalphSteinhagen) [reply include/port.hpp:247]: done.

File: include/node.hpp [hunk 3]
```diff
@@ -243,6 +237,36 @@ class node : protected std::tuple<Arguments...> {
...
     }
 
+    [[nodiscard]] constexpr bool
+    input_tags_present() const noexcept {
+        return _input_tags_present;
+    };
+
+    [[nodiscard]] constexpr bool
+    acknowledge_input_tags() noexcept {
+        if (_input_tags_present) {
```
[the remark this answers is not in the document]
Comment (RalphSteinhagen) [reply include/node.hpp:247]: the function returns 'true' only during the first call... further calls will always return 'false'. This indicates if the tags have already been acknowledged.

File: include/node.hpp [hunk 4]
```diff
@@ -401,13 +432,67 @@ class node : protected std::tuple<Arguments...> {
...
+                                merged_tag_map.insert(map.begin(), map.end());
+                            }
+                            input_port.tagReader().consume(1);
+                        }
+                    },
+                    input_ports(&self()));
+
+            if constexpr (tag_policy == tag_propagation_policy_t::TPP_ALL_TO_ALL) {
+                // N.B. ranges omitted because of missing Clang/Emscripten support
+                std::for_each(_tags_at_output.begin(), _tags_at_output.end(), [](tag_t::map_type &tag) { tag.clear(); });
```
[the remark this answers is not in the document]
Comment (RalphSteinhagen) [reply include/node.hpp:463]: replaced with copy.

File: include/port.hpp [hunk 5]
```diff
@@ -73,51 +107,55 @@ class port {
...
+    [[nodiscard]] std::pair<void *, void *>
     writer_handler_internal() noexcept {
         static_assert(IS_OUTPUT, "only to be used with output ports");
-        return static_cast<void *>(std::addressof(_ioHandler));
+        return { static_cast<void *>(std::addressof(_ioHandler)), static_cast<void *>(std::addressof(_tagIoHandler)) };
     }
 
     [[nodiscard]] bool
-    update_reader_internal(void *buffer_writer_handler_other) noexcept {
+    update_reader_internal(std::pair<void *, void *> buffer_writer_handler_other) noexcept {
```
[the remark this answers is not in the document]
Comment (RalphSteinhagen) [reply include/port.hpp:126]: done

File: include/node.hpp [hunk 6]
```diff
@@ -401,13 +432,67 @@ class node : protected std::tuple<Arguments...> {
...
+            //meta::tuple_for_each([&port_id, this](auto &output_port) noexcept { publish_tag2(output_port, _tags_at_output[port_id++]); }, output_ports(&self()));
+            meta::tuple_for_each(
+                    [&port_id, this](auto &output_port) noexcept {
+                        if (_tags_at_output[port_id].empty()) {
+                            port_id++;
+                            return;
+                        }
+                        auto data           = output_port.tagWriter().reserve_output_range(1);
+                        data[port_id].index = output_port.streamWriter().position();
+                        data[port_id].map.clear();
```
[the remark this answers is not in the document]
Comment (RalphSteinhagen) [reply include/node.hpp:484]: done.

File: include/graph.hpp [hunk 7]
```diff
@@ -122,7 +122,7 @@ class dynamic_port {
                         requires { arg.writer_handler_internal(); }, "'private void* writer_handler_internal()' not implemented");
             } else {
                 static_assert(
-                        requires { arg.update_reader_internal(std::declval<void *>()); }, "'private bool update_reader_internal(void* buffer)' not implemented");
+                        requires { arg.update_reader_internal(std::declval<std::pair<void *, void *>>()); }, "'private bool update_reader_internal(void* buffer)' not implemented");
```
[the remark this answers is not in the document]
Comment (RalphSteinhagen) [reply include/graph.hpp]: done

**Review by ivan-cukic — APPROVED** [review]

**Pull request merged**